lib/cpp/src/thrift/server/TNonblockingServer.cpp

```diff
@@ -547,12 +553,18 @@ void TNonblockingServer::TConnection::transition() {
   case APP_READ_REQUEST:
     // We are done reading the request, package the read buffer into transport
     // and get back some data from the dispatch function
-    inputTransport_->resetBuffer(readBuffer_, readBufferPos_);
-    outputTransport_->resetBuffer();
-    // Prepend four bytes of blank space to the buffer so we can
-    // write the frame size there later.
-    outputTransport_->getWritePtr(4);
-    outputTransport_->wroteBytes(4);
+    if (!server_->getOutputProtocolFactory()) {
+      inputTransport_->resetBuffer(readBuffer_, readBufferPos_);
+      outputTransport_->resetBuffer();
+    } else {
+      inputTransport_->resetBuffer(readBuffer_ + 4, readBufferPos_ - 4);
```

Comment:
THeader expects the first four bytes of framing to still be in the message.  Nonblocking server previously always stripped it off